R/class-workbook.R

```diff
@@ -5400,6 +5728,281 @@ wbWorkbook <- R6::R6Class(
+          if (length(formula) == 2L) {
+            negColour <- formula[[1]]
+            posColour <- formula[[2]]
+          } else {
+            posColour <- formula
+            negColour <- "FFFF0000"
+          }
+
+          extLst <- self$worksheets[[sheet]]$extLst
+
+          guid <- stri_join(
+            "F7189283-14F7-4DE0-9601-54DE9DB",
+            40000L + length(xml_node(extLst, "ext", "x14:conditionalFormattings", "x14:conditionalFormatting"))
+          )
+
+          newExtLst <- gen_databar_extlst(
+            guid      = guid,
+            sqref     = sqref,
+            posColour = posColour,
+            negColour = negColour,
+            values    = values,
+            border    = params$border,
+            gradient  = params$gradient
+          )
+
+          # check if any extLst availaible
+          if (length(extLst) == 0) {
+            self$worksheets[[sheet]]$extLst <- newExtLst
+          } else if (length(xml_node(extLst, "ext", "x14:conditionalFormattings")) == 0) {
+            # extLst is available, has no conditionalFormattings
+            extLst <- xml_add_child(extLst,
+                                    xml_node(newExtLst, "ext", "x14:conditionalFormattings"))
+            self$worksheets[[sheet]]$extLst <- extLst
+          } else {
+            # extLst is available, has conditionalFormattings
+            extLst <- xml_add_child(extLst,
+                                    xml_node(newExtLst, "ext", "x14:conditionalFormattings", "x14:conditionalFormatting"),
+                                    level = "x14:conditionalFormattings")
+            self$worksheets[[sheet]]$extLst <- extLst
+          }
+
+
+
+          if (is.null(values)) {
+            sprintf(
+              '<cfRule type="dataBar" priority="1"><dataBar showValue="%s">
+                          <cfvo type="min"/><cfvo type="max"/>
+                          <color rgb="%s"/>
+                          </dataBar>
+                          <extLst><ext uri="{B025F937-C7B1-47D3-B67F-A62EFF666E3E}" xmlns:x14="http://schemas.microsoft.com/office/spreadsheetml/2009/9/main"><x14:id>{%s}</x14:id></ext>
+                        </extLst></cfRule>',
+              params$showValue,
+              posColour,
+              guid
+            )
+          } else {
+            sprintf(
+              '<cfRule type="dataBar" priority="1"><dataBar showValue="%s">
+                            <cfvo type="num" val="%s"/><cfvo type="num" val="%s"/>
+                            <color rgb="%s"/>
+                            </dataBar>
+                            <extLst><ext uri="{B025F937-C7B1-47D3-B67F-A62EFF666E3E}" xmlns:x14="http://schemas.microsoft.com/office/spreadsheetml/2009/9/main">
+                            <x14:id>{%s}</x14:id></ext></extLst></cfRule>',
+              params$showValue,
+              values[[1]],
+              values[[2]],
+              posColour,
+              guid
+            )
+          }
```

Comment:
If I'm not mistaken, this could be shortened. If the value part `<cfvo type="num" val="%s"/><cfvo type="num" val="%s"/>` is the only difference, we could create it only if `!is.null(values)` and sprintf this string or otherwise just `""`. But I haven't touched it previously, we might as well ignore it for the future.